scipy/_lib/_util.py

```diff
@@ -1111,11 +1111,38 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
 "core" shapes. However, array argument(s) of this function may have additional
 "batch" dimensions prepended to the core shape. In this case, the array is treated
 as a batch of lower-dimensional slices; see :ref:`linalg_batch` for details.
-Note that calls with zero-size batches are unsupported and will raise a ``ValueError``.
 """
 
 
-def _apply_over_batch(*argdefs):
+def output_from_signature(arrays, batch_shape, core_shapes, signature):
+    dtype = np.result_type(*arrays)
+
+    # parse more efficiently with regex?
+    inputs, outputs = signature.split("->")
+    inputs, outputs = inputs[1:-1].split("),("), outputs[1:-1].split("),(")
+    input_dim_to_letter = {}
+    for i, input in enumerate(inputs):
+        for j, l in enumerate(input.split(",")):
+            input_dim_to_letter[(i, j)] = l
+
+    letter_to_length = {}
```

Comment:
This part gathers the numerical length corresponding with each variable in the NEP 5 signature. For instance, if the arrays are of shapes `(3, 4), (5,)`, the dictionary would be `{'m': 3, 'n': 4, 'p': 5}`. Perhaps this could all be done in the same loop in which the signature is processed. This started as a proof of concept and turned into a PR without refactoring. Feel free to make this simpler/more efficient.